### ISSUE TITLE
Keypoints match inaccuracies

### ISSUE THREAD

**Issue by Just-ZJJ** [hero-slam/HERO-SLAM#1]
Dear Author,

Hello. I have attempted to visualize the keypoints matching section of your code, but the results obtained were not satisfactory. The code I added is as follows:
`
  
        '''
        select samples from the image using superpoint heatmap
        '''
  def select_samples_from_sp(self, H, W, samples, rgb, batch=None):

        assert H == rgb.shape[1]
        assert W == rgb.shape[2]

        img = torch.tensor(rgb.permute(0, 3, 1, 2), dtype=torch.float).to(self.device)
        feats, desc_all = self.sp_extractor({'image': img})

        last_img = torch.tensor(self.last_image.permute(0, 3, 1, 2), dtype=torch.float).to(self.device)

        # match last_image and current_image
        data = {'image0': img, 'image1': last_img}
        pred = {**{k + '0': v for k, v in feats.items()},
                **{k + '1': v for k, v in self.last_image_kp.items()},
                **data}
        pred = {**pred, **self.lg_matcher(pred)}
        pred = {k: v.to(self.device).detach()[0] if
        isinstance(v, torch.Tensor) else v for k, v in pred.items()}

        if 'cuda' in str(self.device):
            torch.cuda.empty_cache()

        # create match indices
        matches0, mscores0 = pred['matches0'], pred['matching_scores0']

        valid = matches0 > -1
        matches = torch.stack([torch.where(valid)[0], matches0[valid]], -1)
        pred = {**pred, 'matches': matches, 'matching_scores': mscores0[valid]}

        mscores, indices = mscores0[valid].sort(dim=0, descending=True)
        mscores = mscores[:500]
        indices = indices[:500]
        matches = matches[indices]

        kpts0, kpts1 = pred['keypoints0'], pred['keypoints1']
        m_kpts0, m_kpts1 = kpts0[matches[..., 0]], kpts1[matches[..., 1]]
        m_kpts0 = m_kpts0.detach().cpu().numpy()
        m_kpts1 = m_kpts1.detach().cpu().numpy()
        print("m_kpts0 = ", m_kpts0.shape)
        print("m_kpts1 = ", m_kpts1.shape)

        kpts0 = torch.tensor(np.array(m_kpts0)).to(self.device)
        kpts1 = torch.tensor(np.array(m_kpts1)).to(self.device)

        def visualize_matches(img0, img1, kpts0, kpts1, matches):
            
            img0_np = ((img0*255.0).squeeze().permute(1, 2, 0).cpu().numpy()).astype(np.uint8)
            img1_np = ((img1*255.0).squeeze().permute(1, 2, 0).cpu().numpy()).astype(np.uint8)

            h_max = max(img0_np.shape[0], img1_np.shape[0])
            w_total = img0_np.shape[1] + img1_np.shape[1]
            canvas = np.zeros((h_max, w_total, 3), dtype=np.uint8)
            canvas[:img0_np.shape[0], :img0_np.shape[1]] = img0_np
            canvas[:img1_np.shape[0], img0_np.shape[1]:] = img1_np

            for match in matches:
                if match[0] < len(kpts0) and match[1] < len(kpts1):
                    pt0 = tuple(map(int, kpts0[match[0]]))
                    pt1 = tuple(map(int, kpts1[match[1]]))
                    pt1_shifted = (pt1[0] + img0_np.shape[1], pt1[1])

                    color = tuple(map(int, np.random.randint(0, 255, size=3)))
                    cv2.line(canvas, pt0, pt1_shifted, color, 1, cv2.LINE_AA)
                    cv2.circle(canvas, pt0, 2, color, -1)
                    cv2.circle(canvas, pt1_shifted, 2, color, -1)

            plt.figure(figsize=(20, 10))
            plt.imshow(cv2.cvtColor(canvas, cv2.COLOR_BGR2RGB))
            plt.axis('off')
            plt.show()

        visualize_matches(img, last_img, m_kpts0, m_kpts1, matches)

        return kpts0, kpts1, feats, mscores, desc_all
![20241219111626](https://github.com/user-attachments/assets/f2441212-be79-4c12-b500-7d7699cf96ae)

`

**Comment by xinzhe11** [comment]
I will review the code and respond as soon as possible.

**Comment by xinzhe11** [comment]
@Justzhongjj
the output of select_samples_from_sp is correct, m_kpts0 and m_kpts1 area already sampled by match results.
you should not resample again, this leads to wrong matching results.

pt0 = tuple(map(int, kpts0[match[0]]))
pt1 = tuple(map(int, kpts1[match[1]]))

**Comment by Just-ZJJ** [comment]
@xinzhe11 Thanks for the answer, I got the exact feature point matching result based on your opinion.
![57ff746f-3c05-4a26-b0c6-dc421092db78](https://github.com/user-attachments/assets/bc1e3009-128f-4f02-a213-5936f963ec5c)
